### PR TITLE
Redirrect stderr from compiler into the tcl console (close #302)

### LINKED ISSUE #302: stderr shows in the terminal while GUI mode
It should show in the GUI console instead (in red).

alain@alain-xps:~/FOEDAG$ ./dbuild/bin/foedag --compiler openfpga --script tests/Testcases/raygentop/raygentop.tcl 
% Error 1: 
Type: Routing
File: /home/alain/RapidSilicon/Raptor/OpenFPGA_RS/vpr/src/base/vpr_api.cpp
Line: 801
Message: Fixed channel width must be specified when loading routing (was -1)
Error 1: 
Type: Routing
File: /home/alain/RapidSilicon/Raptor/OpenFPGA_RS/vpr/src/base/vpr_api.cpp
Line: 801
Message: Fixed channel width must be specified when loading routing (was -1)
done!
QObject::~QObject: Timers cannot be stopped from another thread
Segmentation fault (core dumped)

### PR DESCRIPTION
**Pull request by KochynVolodymyr**
Close #302

### CODE REVIEW

File: tests/tclutils/popenRWE.cpp [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  rc = pipe(in);
+  if (rc < 0) goto error_in;
+
+  rc = pipe(out);
+  if (rc < 0) goto error_out;
+
+  rc = pipe(err);
+  if (rc < 0) goto error_err;
+
+  pid = fork();
```
Comment (alain-rs) [inline tests/tclutils/popenRWE.cpp:35]: Absolutely no Fork!
Please remove this component from the source tree. Forks are a terrible idea as they double the process memory size and and not portable.
Please revert back to the Pipe as I coded it.

File: src/Compiler/Compiler.cpp [hunk 2]
```diff
@@ -926,21 +928,33 @@ bool Compiler::ExecuteAndMonitorSystemCommand(const std::string& command) {
...
-  if ((ptr = popen(cmd, "r")) != nullptr) {
-    while (fgets(buf, BUFSIZ, ptr) != nullptr) {
-      if (m_stop == true) {
-        break;
-      }
-      (*m_out) << buf << std::flush;
-    }
-    pclose(ptr);
+  int rwepipe[3];
+  int pid = popenRWE(rwepipe, cmd);
```
Comment (alain-rs) [inline src/Compiler/Compiler.cpp:936]: Do not use this component that uses fork

File: src/Compiler/Compiler.cpp [hunk 3]
```diff
@@ -926,21 +928,33 @@ bool Compiler::ExecuteAndMonitorSystemCommand(const std::string& command) {
...
+  if (pid == -1) return false;
+
+  out = fdopen(rwepipe[1], "r");
+  err = fdopen(rwepipe[2], "r");
+  if (!out || !err) return false;
+
+  char *out_ptr, *error_ptr;
+  while (((out_ptr = fgets(buf, BUFSIZ, out)) != nullptr) ||
+         ((error_ptr = fgets(errbuf, BUFSIZ, err)) != nullptr)) {
```
Comment (alain-rs) [inline src/Compiler/Compiler.cpp:945]: Simply add this err pulling to the original code

Comment (KochynVolodymyr) [reply src/Compiler/Compiler.cpp:945]: I need open file descriptor for stderr? how to do it with old approach?

**Comment by alaindargelas** [comment]
Look for a solution that does not involve fork!

File: tests/tclutils/popenRWE.cpp (hunk 1 above)
Comment (KochynVolodymyr) [reply tests/tclutils/popenRWE.cpp:35]: From the **popen()** documentation :
The popen() function opens a process by creating a pipe, forking,
       and invoking the shell.

Comment (KochynVolodymyr) [reply tests/tclutils/popenRWE.cpp:35]: this approach is the same as it was before but with error handling

**Pull request closed without merging**